### PR TITLE
Accelerate MaxPool

### PR DESCRIPTION
**Pull request by RandySheriffH**
Address issue: https://github.com/microsoft/onnxruntime/issues/3548
Reply on a FIFO queue to accelerate MaxPool on higher pool size.

Perf comparison in **seconds**:
Tensor Shape | BoostMaxPool | Master | Pool Size
------ | ------ | ------ | ------ 
**100\*100\*40960**| **5.07** |**45.89**|**128**
**10\*10\*2048\*2048**| **10.24**|**1744.01**|**64\*64**
**10\*10\*100\*100\*100**| **7.29**|**36.59**|**32\*32\*32**

Add UT models by:
https://github.com/onnx/onnx/pull/3075

### CODE REVIEW

File: onnxruntime/core/providers/cpu/nn/pool.cc [hunk 1]
```diff
@@ -224,6 +224,67 @@ Status MaxPoolV8::ComputeImpl(OpKernelContext* context) const {
...
+  std::vector<int64_t> output_dims = pool_attrs_.SetOutputSize(x_shape, x_shape[1], &pads);
+  const auto& dilations = pool_attrs_.dilations;
+  const auto& pool_size = pool_attrs_.kernel_shape;
+
+  if (dilations[0] != stride_h() ||
+      pool_size.size() > 1 && dilations[1] != stride_w() ||
+      pool_size.size() > 2 && dilations[2] != stride_d() ||
+      context->Output(1, output_dims)) {
+
+    return Status(common::ONNXRUNTIME, common::FAIL);
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/nn/pool.cc:243]: error message along with the fail status ?

File: onnxruntime/core/providers/cpu/nn/pool.cc [hunk 2]
```diff
@@ -224,6 +224,67 @@ Status MaxPoolV8::ComputeImpl(OpKernelContext* context) const {
...
+                                   {X_data, Y_data,
+                                    x_shape[2], x_shape[3], x_shape[4],
+                                    output_dims[2], output_dims[3], output_dims[4],
+                                    pads[0], pads[3], pads[1], pads[4], pads[2], pads[5],
+                                    stride_h(), stride_w(), stride_d(),
+                                    pool_size[0], pool_size[1], pool_size[2]});
+      break;
+    }
+    default:
+      return Status(ONNXRUNTIME, common::FAIL);
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/nn/pool.cc:283]: error message saying > 3D pooling not supported ?

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 3]
```diff
@@ -395,4 +395,309 @@ struct MaxPool3DTask {
...
+  int64_t pool_size;
+
+  TensorOpCost Cost() {
+    auto loop_count = static_cast<double>(x_h);
+    return TensorOpCost{loop_count, loop_count, loop_count};
+  }
+
+  void operator()(std::ptrdiff_t begin, std::ptrdiff_t end) const {
+#ifdef _OPENMP
+#pragma omp parallel for
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:535]: AFAIK- individual operator code shouldn't have OPENMP pragmas and they should use the "abstraction" ThreadPool class to parallelize shards of work - https://github.com/microsoft/onnxruntime/blob/master/docs/NotesOnThreading.md

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 4]
```diff
@@ -395,4 +395,309 @@ struct MaxPool3DTask {
...
+  int64_t pool_w;
+
+  TensorOpCost Cost() {
+    auto loop_count = static_cast<double>(x_h * x_w);
+    return TensorOpCost{loop_count, loop_count, loop_count};
+  }
+
+  void operator()(std::ptrdiff_t begin, std::ptrdiff_t end) const {
+#ifdef _OPENMP
+#pragma omp parallel for
```
Comment (hariharans29) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:582]: same comment as above

File: onnxruntime/core/providers/cpu/nn/pool.cc (hunk 1 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cpu/nn/pool.cc:243]: The returned status will not be reported or consumed anywhere else beyond its caller.

File: onnxruntime/core/providers/cpu/nn/pool.cc (hunk 2 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cpu/nn/pool.cc:283]: Same as above.

File: onnxruntime/core/providers/cpu/nn/pool_functors.h (hunk 3 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cpu/nn/pool_functors.h]: Removed.

File: onnxruntime/core/providers/cpu/nn/pool_functors.h (hunk 4 above)
Comment (RandySheriffH) [reply onnxruntime/core/providers/cpu/nn/pool_functors.h]: Removed.

File: onnxruntime/core/providers/cpu/nn/pool.cc (hunk 1 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/nn/pool.cc:243]: The error msg should trickle all the way up the chain. Is that not the case?

Comment (RandySheriffH) [reply onnxruntime/core/providers/cpu/nn/pool.cc:243]: No it is not, let me return a bool to relieve the confusion.

Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/nn/pool.cc:243]: Let's fix the error msg propagation then. How would a user know if something went wrong.

Comment (RandySheriffH) [reply onnxruntime/core/providers/cpu/nn/pool.cc]: Done.

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 5]
```diff
@@ -395,4 +395,274 @@ struct MaxPool3DTask {
...
+  int64_t pad_h_front;
+  int64_t pad_h_back;
+  int64_t pad_w_front;
+  int64_t pad_w_back;
+  int64_t dilation_h;
+  int64_t dilation_w;
+  int64_t pool_h;
+  int64_t pool_w;
+
+  TensorOpCost Cost() {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:559]: >) { [](start = 20, length = 3)

const?

File: onnxruntime/core/providers/cpu/nn/pool_functors.h [hunk 6]
```diff
@@ -395,4 +395,274 @@ struct MaxPool3DTask {
...
+  const T* X_data;
+  T* Y_data;
+  int64_t x_h;
+  int64_t y_h;
+  int64_t pad_front;
+  int64_t pad_back;
+  int64_t dilation;
+  int64_t pool_size;
+
+  TensorOpCost Cost() {
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/nn/pool_functors.h:520]: > { [](start = 21, length = 2)

const?

**Comment by snnn** [comment]
How is it going?

**Comment by snnn** [comment]
I will close it now.

**Pull request closed without merging**